File: cpp/scene/Skybox.cpp
#include "Skybox.hpp"
#include "Camera.hpp"
#include "renderer/Backend.hpp"
#include "renderer/Shader.hpp"
#include "settings/Settings.hpp"

#include <glm/glm.hpp>
#include <glm/gtc/matrix_transform.hpp>

constexpr float Skybox::vertices[];

void Skybox::setup(Backend &b) {
  backend = &b;

  b.createSkyboxMesh(vertices, sizeof(vertices), vao, vbo);

  texture = b.loadCubemap({
      "textures/skybox/right.png",
      "textures/skybox/left.png",
      "textures/skybox/top.png",
      "textures/skybox/bottom.png",
      "textures/skybox/front.png",
      "textures/skybox/back.png",
  });
}

void Skybox::destroy() {
  if (!backend)
    return;
  if (vao)
    backend->destroySkyboxMesh(vao, vbo);
  if (texture)
    backend->destroyTexture(texture);
}

void Skybox::render(const Shader &shader, const Camera &cam) const {
  backend->setDepthFunc(true);
  shader.use();

  glm::mat4 view =
      glm::mat4(glm::mat3(glm::lookAt(cam.pos, cam.pos + cam.front, cam.up)));
  glm::mat4 proj = glm::perspective(glm::radians(cam.fov),
                                    Settings::aspectRatio(), 0.1f, 100.0f);

  shader.setMat4("view", view);
  shader.setMat4("projection", proj);
  shader.setInt("skybox", 0);

  backend->bindCubemap(0, texture);
  backend->drawSkybox(vao);

  backend->setDepthFunc(false);
}

File: go_deprecated/scene/skybox.go
package scene

import (
	"github.com/go-gl/mathgl/mgl32"

	"github.com/Zephyr75/overdrive/renderer"
	"github.com/Zephyr75/overdrive/settings"
)

type Skybox struct {
	mesh    renderer.MeshHandle
	Texture renderer.TextureHandle
}

func (s *Skybox) setup(b renderer.Backend) {
	vertices := []float32{
		// positions
		-1.0, 1.0, -1.0,
		-1.0, -1.0, -1.0,
		1.0, -1.0, -1.0,
		1.0, -1.0, -1.0,
		1.0, 1.0, -1.0,
		-1.0, 1.0, -1.0,

		-1.0, -1.0, 1.0,
		-1.0, -1.0, -1.0,
		-1.0, 1.0, -1.0,
		-1.0, 1.0, -1.0,
		-1.0, 1.0, 1.0,
		-1.0, -1.0, 1.0,

		1.0, -1.0, -1.0,
		1.0, -1.0, 1.0,
		1.0, 1.0, 1.0,
		1.0, 1.0, 1.0,
		1.0, 1.0, -1.0,
		1.0, -1.0, -1.0,

		-1.0, -1.0, 1.0,
		-1.0, 1.0, 1.0,
		1.0, 1.0, 1.0,
		1.0, 1.0, 1.0,
		1.0, -1.0, 1.0,
		-1.0, -1.0, 1.0,

		-1.0, 1.0, -1.0,
		1.0, 1.0, -1.0,
		1.0, 1.0, 1.0,
		1.0, 1.0, 1.0,
		-1.0, 1.0, 1.0,
		-1.0, 1.0, -1.0,

		-1.0, -1.0, -1.0,
		-1.0, -1.0, 1.0,
		1.0, -1.0, -1.0,
		1.0, -1.0, -1.0,
		-1.0, -1.0, 1.0,
		1.0, -1.0, 1.0,
	}

	s.mesh = b.CreateSkyboxMesh(vertices)
	tex, err := b.LoadCubemap([6]string{
		"./textures/skybox/right.png",
		"./textures/skybox/left.png",
		"./textures/skybox/top.png",
		"./textures/skybox/bottom.png",
		"./textures/skybox/front.png",
		"./textures/skybox/back.png",
	})
	if err != nil {
		println("Error loading skybox:", err.Error())
	}
	s.Texture = tex
}

func (s *Scene) RenderSkybox(shader renderer.ShaderHandle, u *renderer.Uniforms) {
	// View with the translation stripped, so the skybox follows the camera.
	view := mgl32.LookAtV(s.Cam.Pos, s.Cam.Pos.Add(s.Cam.Front), s.Cam.Up)
	u.View = view.Mat3().Mat4()
	u.Projection = mgl32.Perspective(mgl32.DegToRad(s.Cam.Fov),
		float32(settings.WindowWidth)/float32(settings.WindowHeight), 0.1, 100.0)
	u.TexSkybox = s.Skybox.Texture

	s.backend.SetDepthFunc(true) // depth <= 1.0 passes at the far plane
	s.backend.DrawSkybox(shader, s.Skybox.mesh, u)
	s.backend.SetDepthFunc(false)
}
